xds/src/main/java/io/grpc/xds/internal/security/trust/XdsX509TrustManager.java

```diff
@@ -245,19 +258,37 @@ public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngi
       sslParams.setEndpointIdentificationAlgorithm("");
       sslEngine.setSSLParameters(sslParams);
     }
-    delegate.checkServerTrusted(chain, authType, sslEngine);
+    chooseDelegate(chain).checkServerTrusted(chain, authType, sslEngine);
     verifySubjectAltNameInChain(chain);
   }
 
   @Override
   public void checkServerTrusted(X509Certificate[] chain, String authType)
       throws CertificateException {
-    delegate.checkServerTrusted(chain, authType);
+    chooseDelegate(chain).checkServerTrusted(chain, authType);
     verifySubjectAltNameInChain(chain);
   }
 
+  private X509ExtendedTrustManager chooseDelegate(X509Certificate[] chain)
+      throws CertificateException {
+    if (delegates != null) {
+      Optional<SpiffeUtil.SpiffeId> spiffeId = SpiffeUtil.extractSpiffeId(chain);
+      if (!spiffeId.isPresent()) {
+        throw new CertificateException("Can't extract valid SPIFFE ID from the chain");
```

Comment:
Suggested edit for clarity: "Failed to extract SPIFFE ID from peer leaf certificate."